### PR TITLE
OrbitControls: Support Ctrl-drag as right-drag alternative.

### LINKED ISSUE #13970: OrbitControls should support Ctrl-drag (etc.)
##### Description of the problem

WebKit is making use of OrbitControls for a [visualization](https://webkit.org/blog/8262/visualizing-layers-in-web-inspector/) in Web Inspector, and we realized that it would be convenient—particularly with a trackpad—if one could pan with Ctrl-drag as an alternative to right-drag.

I don't know whether people would be receptive to this as default behavior, but if not, it would still be great if we could build upon `mouseButtons` such that consumers can make this configuration from the outside.

Either way, as long as the functionality is desired, I'd be happy to submit a PR. :smile:

##### Three.js version

- [x] r92

##### Browser

- [x] All of them

##### OS

- [x] All of them

### PR DESCRIPTION
**Pull request by rkirsling**
Resolves #13970.

Notes:

1. As mentioned in the corresponding issue, whether this is appropriate default behavior or should simply be configurable by users is your call.

2. Since this change requires introducing support not only for Ctrl (_and perhaps Meta/Alt/Shift_) but also for associating multiple triggers with a single action, I believe it's simplest to use lambda predicates (rather than, say, arrays of objects, which we'd need a function call to filter through anyway).

3. Admittedly `ORBIT(event)` looks a bit odd due to the enum casing, but I feel like this is preferable to juggling a possibly-undefined function `isOrbit` alongside `ORBIT`.

### CODE REVIEW

File: examples/js/controls/OrbitControls.js [hunk 1]
```diff
@@ -675,37 +689,29 @@ THREE.OrbitControls = function ( object, domElement ) {
 
 		event.preventDefault();
 
-		switch ( event.button ) {
+		if ( typeof scope.mouseButtons.ORBIT === 'function' ? scope.mouseButtons.ORBIT(event) : event.button === scope.mouseButtons.ORBIT ) {
```
Comment (BlueRaja) [inline examples/js/controls/OrbitControls.js]: Pull this logic into a function so you don't need to repeat it so often

**Comment by WestLangley** [comment]
Without passing judgement on the advisability of this feature, I would consider a different implementation:
```javascript
function onMouseDown( event ) {

	if ( scope.enabled === false ) return;

	event.preventDefault();

	switch ( scope.mouseMaping( event ) ) {

		case scope.mouseButtons.ORBIT:
	. . .
```
Then
```javascript
this.mouseButtons = { ORBIT: 1001, ZOOM: 1002, PAN: 1003 }; // arbitrary values

this.mouseMaping = function ( event ) {

	if ( event.button === THREE.MOUSE.LEFT && ! event.ctrlKey ) return scope.mouseButtons.ORBIT;

	if ( event.button === THREE.MOUSE.MIDDLE ) return scope.mouseButtons.ZOOM;

	if ( event.button === THREE.MOUSE.RIGHT || ( event.button === THREE.MOUSE.LEFT && event.ctrlKey ) ) return scope.mouseButtons.PAN;

	return null;

}
```
I would expect there is even a better way...

**Comment by rkirsling** [comment]
Sure, another implementation I'd been considering is similar to that. It preserves the `switch` but calls a new `scope.getStateForMouseDown( event )` and switches on the new `STATE` enum value.

Reasoning:

- If we're concerned about backwards-compatibility, then we can't change the type of `mouseButtons` in a destructive way, and so we need the new function to _use_ the buttons and not return them.

- If we preserve the `switch`, we're going to have to walk two three-tiered conditionals (three `if`s and three `case`s) in a row to determine whether a `mousedown` event is a `PAN` action, but switching on the new `STATE` lessens redundancy and allows for a bit of simplification in `onMouseDown`.

I'll update the PR to go with that approach—hope it's more satisfying to you. :smile:

**Comment by WestLangley** [comment]
Live: https://rawgit.com/rkirsling/three.js/orbitcontrols-ctrldrag/examples/misc_controls_orbit.html

**Comment by WestLangley** [comment]
I like this feature. Let's see if we can get some feedback from others.

Sorry, I'm still not sure about the implementation details, though...

**Comment by rkirsling** [comment]
@WestLangley Thanks! Is there anything I can/should be doing to elicit such feedback?

**Comment by WestLangley** [comment]
Maybe @moroine has some suggestions...

**Comment by moroine** [comment]
Nice feature, about the implementation I would go with bit-mask but I might be too complicated to use for beginners.

Having something like 

```javascript
THREE.MODIFIER_KEYS = {  CTRL: 1, ALT: 2, SHIFT: 4, META: 8 };

this.mouseButtons = { 
  ORBIT: {
     0: THREE.MOUSE.LEFT, // default one
     1: 0 // No binding ==> disabled
  },
  ZOOM: THREE.MOUSE.MIDDLE,
  PAN: {
    0: THREE.MOUSE.RIGHT,
    1: THREE.MOUSE.LEFT, 
  }
};
```

We could ensure the backward compatibility, but I'm not even convinced myself :disappointed: 

About the feature itself, I've face a limitation. When you start rotating and press CTRL it doesn't switch to Drag mode until the next mouseup event.

**Comment by rkirsling** [comment]
Assuming we allow users to configure this (instead of just silently equating Ctrl-drag to right-drag internally, which I suppose is another valid possibility?), I agree that a bitmask approach seems like an overly expert-oriented API. :sweat:

Regarding the limitation you described, I would personally consider this to be expected behavior. One major precedent for this is [Google Earth](https://earth.google.com/), in which Shift-drag changes the rotation (but only at the time of the mousedown event).

**Comment by moroine** [comment]
On the other hand your method `getStateForMouseDown` can be "public option" used to configure the behavior.

This would replace `mouseButtons`, we could let mouseButtons in order to do easy configuration and let the ability to replace  `getStateForMouseDown` for more advanced configurations.

**Comment by rkirsling** [comment]
Works for me! :smile:

**Comment by WestLangley** [comment]
@mrdoob I think this is worthy of being merged.

**Pull request merged**